### PR TITLE
Profile configurations for InnerOuterPersistent scheduler in python frontend

### PR DESCRIPTION
**Pull request by rdspring1**
# Summary
This PR explores auto-tuning a `LayerNormBackward` fusion using the `InnerOuterPersistent` scheduler in the python-frontend.

- Create `autotune_persistent.py` to test several parameter configurations then apply `DecisionTreeRegressor`
- The selected performance metric is `effective_bandwidth_gbs`. The empirical scheduler selects the configuration that has the highest predicted `effective_bandwidth_gbs`.

# Key differences from approach for `Pointwise` scheduler
- `vectorize_factor`, `thread_per_block_min`, and `thread_per_block_max` are specified before running `computeHeuristics`. These settings are akin to hyper-parameters used to constrain the generated scheduler parameters.
- Create `SchedulerHyperParameters` as an entry in `HeuristicDataCache` to specify these constraints when generating scheduler parameters.

# Details
1. Create `struct SchedulerHyperParameters` in `csrc/scheduler/utils.h`
2. Create `HeuristicDataCacheEntry` in `csrc/scheduler/compile_time_info.h`
3. Modify `computeHeuristics` to use hyper-parameter constraints.
4. Expose `SchedulerHyperParameters` in python frontend.
5. Allow user schedulers to define a `HeuristicDataCache` during scheduling.

* `ScheduleHyperParameters` contains parameters for `vectorize_factor`, `unroll_factor`, `threads_per_block_min`, and `threads_per_block_max`.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
Profile kernel running time (ms) for `LayerNormBackward` kernels.

Batch size: [512, 2048, 8192, 16384]
Hidden size: [1024, 28672] in 256 increments
persistent_compare_empirical_batchsize512
persistent_compare_empirical_batchsize2048
persistent_compare_empirical_batchsize8192
persistent_compare_empirical_batchsize16384

**Comment by rdspring1** [comment]
!build

**Comment by liqiangxl** [comment]
Instead of setting `thread_per_block_min`, what about directly set `thread_per_block `? It allows us to overwrite the internal `thread_per_block_max`, achieving `what we set is what we get`.

File: csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -3659,6 +3690,26 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+            "Attempting to use a SchedOperators Op prior to definition!");
+        UserSchedule* sched = self.fusion_definition->userSchedule();
+        auto scheduler_hyperparameters_entry = HeuristicDataCacheEntry<
+            HeuristicCompileTime::SchedulerHyperParameters>(
+            sched->data_cache.get(), []() {
+              return std::make_unique<
+                  scheduler_utils::SchedulerHyperParameters>(
+                  /*vectorize_factor=*/1,
+                  /*unroll_factor=*/1,
+                  /*threads_per_block=*/1);
```
[the remark this answers is not in the document]
Comment (rdspring1) [reply csrc/python_frontend/python_bindings.cpp]: Perhaps. I just created a single class for now.

**Review by liqiangxl — APPROVED** [review]
LGTM.
Thanks for adding this new feature!

**Comment by rdspring1** [comment]
!build

**Pull request merged**